### PR TITLE
fix(errors): better error code handling

### PR DESCRIPTION
**Pull request by carlhannes**
Improves the error code handling in other projects using enigma.js by specifing a table to lookup when looking for error codes. I've placed the `error-codes.js` in the root to allow it to be imported as `import enigmaErrorCodes from 'enigma.js/error-codes';`

I've searched through the source code and replaced all existing `new Error(` with this new `createEnigmaError` function to streamline this. The only place I've not modified is `error-response-interceptor.js` because, naturally, it sets the error code from the QIX error that it has recieved on the websocket.

`createEnigmaError` does not break existing errors because it only attaches two new things to the existing error with the existing error message:
1. An error code that can be mapped from 'error-codes.js' (Error.code = -5)
2. enigmaError boolean to identify an enigmaError (Error.enigmaError = true) 

## These are the error codes I've specified
```js
export default {
  NOT_CONNECTED: -1, // You're trying to send data on a socket that's not created
  OBJECT_NOT_FOUND: -2, // The object you're trying to fetch does not exist
  EXPECTED_ARRAY_OF_PATCHES: -3, // Unexpected RPC response, expected array of patches
  PATCH_HAS_NO_PARENT: -4, // Patchee is not an object we can patch
  ENTRY_ALREADY_DEFINED: -5, // This entry is already defined with another key
  NO_CONFIG_SUPPLIED: -6, // You need to supply a configuration
  PROMISE_REQUIRED: -7, // There's no promise object available (polyfill required?)
  SCHEMA_STRUCT_TYPE_NOT_FOUND: -8, // The schema struct type you requested does not exist
  SCHEMA_MIXIN_CANT_OVERRIDE_FUNCTION: -9, // Can't override this function
  SCHEMA_MIXIN_EXTEND_NOT_ALLOWED: -10, // Extend is not allowed for this mixin
  SESSION_SUSPENDED: -11, // Session suspended - no interaction allowed
  SESSION_NOT_ATTACHED: -12, // onlyIfAttached supplied, but you got SESSION_CREATED
};
```

### CODE REVIEW

File: src/error.js [hunk 1]
```diff
@@ -0,0 +1,6 @@
+export default function createEnigmaError(code, name) {
```
Comment (egilsster) [inline src/error.js]: Could you move the inline comments to a JSDoc formatted comment above the function?

Comment (egilsster) [reply src/error.js]: Could you also add a spec file for this, that just creates an error with some code / name and verifies it has the correct properties?

Comment (carlhannes) [reply src/error.js]: I'll fix that.... On Monday 😀

**Comment by carlhannes** [comment]
@hrigner @egilsster @stoffeastrom I've updated now with API docs, JSDoc formatted function comments & tests :)

**Pull request merged**